File: 정수빈/LeetCode/easy/path-sum.py
# https://leetcode.com/problems/path-sum/?envType=study-plan-v2&envId=top-interview-150
# 난이도: easy
# 시간: 14분

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        self.answer = False
        def dfs(node, total):
            if node.right == None and node.left == None and total == targetSum:
                self.answer = True
                return 
            
            if node.left:
                dfs(node.left, total + node.left.val)
            if node.right:
                dfs(node.right, total + node.right.val)
        if root:
            dfs(root, root.val)
            
        return self.answer

File: 장세훈/LeetCode/easy/PathSum.java
class PathSum {
    /*
        112. Path Sum
        https://leetcode.com/problems/path-sum/description/?envType=study-plan-v2&envId=top-interview-150

        tree의 root에서 leaf까지의 경로의 합이 targetSum와 같은게 있다면 true 반환
    */
    public static void main(String[] args) {
        new PathSum().hasPathSum(
                new TreeNode(1, new TreeNode(2), null), 1
        );
    }

    boolean answer = false;

    public boolean hasPathSum(TreeNode root, int targetSum) {
        if (root == null) return false;
        dfs(root, targetSum, 0);
        return answer;
    }

    void dfs(TreeNode node, int targetSum, int sum) {
        if (answer) return;
        if (node != null)
            sum += node.val;

        if (node.right == null && node.left == null) {
            if (sum == targetSum)
                answer = true;
            return;
        }

        if(node.left != null) dfs(node.left, targetSum, sum);
        if(node.right != null) dfs(node.right, targetSum, sum);
    }
}